subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestClassSelectionMatcher.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing.filter;
+
+import org.apache.commons.lang.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Excludes as many classes as possible before sending them to {@link org.gradle.api.internal.tasks.testing.TestClassProcessor}
+ * in order to improve test performance. Matching rules are:
+ *
+ * <ul>
+ * <li>If pattern starts with a wildcard (*), we can't exclude any class. E.g., pattern '*testMethod' can match 'testMethod' in any class.</li>
+ * <li>If pattern doesn't starts with a wildcard (*), we can exclude the classes that can't be matched at all. E.g., 'com.gradle.Test' can be excluded by pattern 'org.gradle*'.</li>
+ * <li>Two kinds of special cases without any wildcards are supported: TestClass/TestClass.testMethod can match TestClass in any package.</li>
+ * </ul>
+ */
+public class TestClassSelectionMatcher {
+    private final List<TestClassPattern> patterns = new ArrayList<TestClassPattern>();
+
+    public TestClassSelectionMatcher(Collection<String> includedTests, Collection<String> includedTestsCommandLine) {
+        for (String pattern : includedTests) {
+            patterns.add(new TestClassPattern(pattern));
+        }
+        for (String pattern : includedTestsCommandLine) {
+            patterns.add(new TestClassPattern(pattern));
+        }
+    }
+
+    public boolean maybeMatchClass(String fullQualifiedName) {
+        if (patterns.isEmpty()) {
+            return true;
+        }
+        String simpleName = getSimpleName(fullQualifiedName);
+        for (TestClassPattern pattern : patterns) {
+            if (pattern.maybeMatchClass(fullQualifiedName, simpleName)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private String getSimpleName(String fullQualifiedName) {
+        String simpleName = StringUtils.substringAfterLast(fullQualifiedName, ".");
+        if ("".equals(simpleName)) {
+            simpleName = fullQualifiedName;
+        }
+        return simpleName;
+    }
+
+    private static class TestClassPattern {
+        private boolean containsWildcard;
```

Comment:
⭕️ You might want to split this into two classes instead of using a boolean here. A `WildCardClassPattern` and a `SimpleClassPattern` or something similar. You can then use a factory method to decide which one to return. That makes each class easier to understand.

Reply:
Also, if you split them into two different classes, you can reuse them in `TestSelectionMatcher`, using two slightly different methods on each class. One could be `matches`, the other could be `mayMatch`.